src/renderer/atlas/AtlasEngine.cpp

```diff
@@ -540,1113 +479,495 @@ CATCH_RETURN()
-    //   To force the destruction, call ID3D11DeviceContext::ClearState (or otherwise ensure
-    //   no views are bound to pipeline state), and then call Flush on the immediate context.
-    if (_r.swapChain && _r.deviceContext)
-    {
-        if (_r.d2dMode)
-        {
-            _r.d2dRenderTarget.reset();
-        }
-        _r.frameLatencyWaitableObject.reset();
-        _r.swapChain.reset();
-        _r.renderTargetView.reset();
-        _r.deviceContext->ClearState();
-        _r.deviceContext->Flush();
-    }
+    TextAnalysisSource analysisSource{ text, textLength };
+    const auto& textFormatAxis = _api.textFormatAxes[static_cast<size_t>(_api.attributes)];
+
+    // We don't read from scale anyways.
+#pragma warning(suppress : 26494) // Variable 'scale' is uninitialized. Always initialize an object (type.5).
+    f32 scale;
+
+    if (textFormatAxis)
+    {
+        THROW_IF_FAILED(_p.systemFontFallback1->MapCharacters(
+            /* analysisSource     */ &analysisSource,
+            /* textPosition       */ 0,
+            /* textLength         */ textLength,
+            /* baseFontCollection */ _p.s->font->fontCollection.get(),
+            /* baseFamilyName     */ _p.s->font->fontName.c_str(),
+            /* fontAxisValues     */ textFormatAxis.data(),
+            /* fontAxisValueCount */ gsl::narrow_cast<u32>(textFormatAxis.size()),
+            /* mappedLength       */ mappedLength,
+            /* scale              */ &scale,
+            /* mappedFontFace     */ reinterpret_cast<IDWriteFontFace5**>(mappedFontFace)));
+    }
+    else
+    {
+        const auto baseWeight = WI_IsFlagSet(_api.attributes, FontRelevantAttributes::Bold) ? DWRITE_FONT_WEIGHT_BOLD : static_cast<DWRITE_FONT_WEIGHT>(_p.s->font->fontWeight);
+        const auto baseStyle = WI_IsFlagSet(_api.attributes, FontRelevantAttributes::Italic) ? DWRITE_FONT_STYLE_ITALIC : DWRITE_FONT_STYLE_NORMAL;
+        wil::com_ptr<IDWriteFont> font;
+
+        THROW_IF_FAILED(_p.systemFontFallback->MapCharacters(
+            /* analysisSource     */ &analysisSource,
+            /* textPosition       */ 0,
+            /* textLength         */ textLength,
+            /* baseFontCollection */ _p.s->font->fontCollection.get(),
+            /* baseFamilyName     */ _p.s->font->fontName.c_str(),
+            /* baseWeight         */ baseWeight,
+            /* baseStyle          */ baseStyle,
+            /* baseStretch        */ DWRITE_FONT_STRETCH_NORMAL,
+            /* mappedLength       */ mappedLength,
+            /* mappedFont         */ font.addressof(),
+            /* scale              */ &scale));
+
+        if (font)
+        {
+            THROW_IF_FAILED(font->CreateFontFace(reinterpret_cast<IDWriteFontFace**>(mappedFontFace)));
+        }
+    }
+
+    // Oh wow! You found a case where scale isn't 1! I tried every font and none
+    // returned something besides 1. I just couldn't figure out why this exists.
+    assert(scale == 1);

```

Comment:
If this assertion fails in a production build, the worst that can happen is that the font size of a fallback font is not quite right. But that doesn't cause any crashes or something. So yeah, I think this doesn't do any harm, but I think it's a good idea to debug-assert on it, since unknown behavior is not good. 😄